File: daemon-rust-archived/src/ingest/stop.rs
//! `POST /hooks/stop` and `POST /hooks/subagent-stop` handler (shared —
//! both events map identically: `session_state::transition` -> `done`,
//! `session_state::timeline_kind` -> `"completion"`, D-05: a finished
//! session stays prominent until acted on or dismissed).

use axum::{extract::State, http::StatusCode, Json};
use serde_json::Value;
use std::sync::Arc;

use crate::{ingest, session_state::HookEvent, AppState, IngestEventRequest};

pub async fn stop(
    State(state): State<Arc<AppState>>,
    Json(payload): Json<Value>,
) -> Result<StatusCode, StatusCode> {
    let session_id = payload
        .get("session_id")
        .and_then(Value::as_str)
        .ok_or(StatusCode::BAD_REQUEST)?
        .to_string();
    let cwd = payload
        .get("cwd")
        .and_then(Value::as_str)
        .map(str::to_string);
    let summary = payload
        .get("last_assistant_message")
        .and_then(Value::as_str)
        .map(|s| ingest::condensed_text(s, 200))
        .unwrap_or_else(|| "Agent turn complete".to_string());

    let request = IngestEventRequest {
        session_id,
        event: HookEvent::Stop,
        cwd,
        notification_type: None,
        tool_name: None,
        timeline_summary: summary,
        payload_json: Some(payload.to_string()),
        is_error: false,
        first_prompt_text: None,
        mark_ended: false,
    };

    ingest::dispatch_ingest_event(&state, request).await
}

File: daemon/src/ingest/stop.ts
/**
 * `POST /hooks/stop` and `POST /hooks/subagent-stop` handler (shared — both
 * events map identically: `sessionState.transition` -> `done`,
 * `sessionState.timelineKind` -> `"completion"`; a finished session stays
 * prominent until acted on or dismissed). Mirrors `daemon/src/ingest/stop.rs`
 * — `routes.ts` registers this SAME handler under both route paths, exactly
 * as `ingest/mod.rs` routes `/hooks/subagent-stop` to `stop::stop`.
 */

import type { FastifyReply, FastifyRequest } from "fastify";
import type { Database as DatabaseType } from "better-sqlite3";

import { condensedText } from "../store.js";
import { dispatchIngestEvent } from "./dispatch.js";
import type { IngestEventRequest } from "./dispatch.js";

export function makeStopHandler(db: DatabaseType) {
  return async (req: FastifyRequest, reply: FastifyReply): Promise<void> => {
    const body = (req.body ?? {}) as Record<string, unknown>;
    const sessionId = typeof body.session_id === "string" ? body.session_id : undefined;
    if (!sessionId) {
      reply.code(400).send();
      return;
    }
    const cwd = typeof body.cwd === "string" ? body.cwd : null;
    const lastMessage =
      typeof body.last_assistant_message === "string" ? body.last_assistant_message : undefined;
    const summary = lastMessage !== undefined ? condensedText(lastMessage, 200) : "Agent turn complete";

    const request: IngestEventRequest = {
      sessionId,
      // Both /hooks/stop and /hooks/subagent-stop map to the same HookEvent
      // — transition()/timelineKind() treat Stop and SubagentStop
      // identically, so the choice here is immaterial to the result
      // (mirrors stop.rs, which always builds HookEvent::Stop regardless
      // of which route invoked it).
      event: "Stop",
      cwd,
      notificationType: null,
      toolName: null,
      timelineSummary: summary,
      payloadJson: JSON.stringify(body),
      isError: false,
      firstPromptText: null,
      markEnded: false,
    };

    dispatchIngestEvent(db, request);
    reply.code(200).send();
  };
}
